src/fiber/fiber.ml

```diff
@@ -392,6 +392,62 @@ module Ivar = struct
       | Empty _ -> None )
 end
 
+module Mvar = struct
+  type 'a t =
+    { writers : ('a * unit K.t) Queue.t
+    ; readers : 'a K.t Queue.t
+    ; mutable value : 'a option
+    }
+
+  let create () =
+    { value = None; writers = Queue.create (); readers = Queue.create () }
+
+  (** [step mvar] makes progress on the passing values to readers/writers. It
+      can detect progress in situations:
+
+      - pending writes when [value] is empty: we can deque this write and make
+        unblock the writer
+
+      - pending reads when [value] is present: we can pass a value to the reader
+        and unblock it
+
+      if [step] is able to make progress, it will call itself recursively to
+      make more progress. That's because running a continuation may make it
+      possible to proceed further. *)
```

Comment:
Actually, I don't understand why this is necessary. If running a continuation allows to make further progress, then it is the operation inside the continuation that should trigger progress. The way `read` and `write` are written is a bit suspicious to me as well, I feel like it could easily starve readers or writers. For instance if you have:

```
fork_and_join
   (fun () ->
       let* x = Mvar.read m in
       printf "reader1: got %d" x;
       Mvar.write m 1)
   (fun () ->
       let* () = Scheduler.yield () in
       let* () = Mvar.write m 2 in
       let* x = Mvar.read m in
       printf "reader2: got %d" x)
```

then `reader2` will get `2` and `reader1` will get stuck. I would expect `reader1` to get `2` since it starter earlier and `reader2` to get `1`.

Reply:
Good fine. I pushed this as a test and attempted to fix this.

I've tried re-implementing mvar in the simplest way possible, but honestly, I'm still not 100% sure it's completely correct. At least my previous tests, and your test seem to pass :/

Reply:
> Actually, I don't understand why this is necessary. If running a continuation allows to make further progress, then it is the operation inside the continuation that should trigger progress

Consider a successful read, it should trigger an attempt to trigger the next write. Otherwise, the fiber waiting on the write will remain blocked. Same goes for successful writes -- those should attempt to unblock readers.

I've pushed a couple of commit that push the progress step into the continuation. It seems a bit clearer.

Reply:
The important part is that a mvar can only make progress on call to `read` or `write` at a time, so there is no need to for any form of recursion: each call to `read` or `write` will do its bit and that's enough. The easiest way to write such functions is to first write a data-structure invariant and reason about it.

To write the invariant, we need to go back the definition of a mvar. I took the definition on [this page](https://www.oreilly.com/library/view/parallel-and-concurrent/9781449335939/ch07.html), adapted it and wrote it in the doc:

```ocaml
  (** A mailbox variable can be thought of as a box that is either empty or
      full. [create ()] creates a new empty box, and [create_full x] creates a
      new full box containing [x].

      [read] removes the value from a full mailbox variable and returns it, but
      blocks if the mvar is currently empty. Symmetrically, [write] puts a value
      into the mvar but blocks if the mvar is already full. *)
```

Given this definition, we can establish the following invariant:

- if the mvar is empty, there can be no writters
- if the mvar is full, there can be be no readers

or put in code:

```ocaml
  let invariant t =
    match t.value with
    | None -> Queue.is_empty t.writers
    | Some _ -> Queue.is_empty t.readers
```

given this invariant, it becomes quite easy to write `read` and `write`. Let's consider `read`:

- if the mvar is empty, we know that there are no writers so we can just queue a reader
- if the mvar is full, there might be some writers:
  - if there are no writers, we can just consume the value and move on
  - if there is at least one writer, we must:
    - replace the current value of the mvar with the one that this writer wants to write
    - notify this writer that the write has completed
    - continue the current fiber with the value that was previously in the mvar

which leads to the following implementation:

```ocaml
  let read t k =
    match t.value with
    | None -> Queue.push t.readers (K.create k)
    | Some v -> (
      match Queue.pop t.writers with
      | None ->
        t.value <- None;
        k v
      | Some (v', w) ->
        t.value <- Some v';
        k v;
        K.run w () )
```

Reply:
Thanks, this makes  it far simpler.